### DISCUSSION TITLE
Can we have this customizations in chakra react select

### DISCUSSION THREAD

**Discussion by howdyAnkit** [csandman/chakra-react-select#37] [Q&A] [answered]
First of all huge thanks for Building this awesome component and library.Just wanted some suggestion if following possibilites are possible in the library.

Can we checkbox the inside the dropdown along with the label
can we have a radio button in the dropdown along with the label
Can we change the dropdown border colour of the dropdownn.

Please find the attached below for references.

![image](https://user-images.githubusercontent.com/40737659/149521744-3107ce32-e9ea-4a29-b0e0-50e10f955a1c.png)

**Comment by csandman** [comment]
Let's start with the easiest question first, the focused border color of the input can be changed using the [`focusBorderColor`](https://github.com/csandman/chakra-react-select#focusbordercolor--default-blue500--errorbordercolor--default-red500) prop, and it can be set to any chakra named color or color string.

As for the rest, it is possible to style a selected option with a checkmark, however its not a checkbox exactly, it's the style of the [selected menu option from chakra](https://chakra-ui.com/docs/overlay/menu#menu-option-groups).  This can be used instead of the default color highlighting by passing [`selectedOptionStyle="check"`](https://github.com/csandman/chakra-react-select#selectedoptionstyle--options-color-check--default-color) and it looks like this:

![image](https://user-images.githubusercontent.com/9214195/149597498-23e7c6b2-3e7a-4bbe-80a6-98af0f4f4514.png)

If you want to change the styles of a selected option by adding your own custom icon, there are two ways you could do it. The first way would be to wrap my `Option` component [like in the react-select docs](https://react-select.com/components#defining-components) and pass it back into the root select. The custom components implemented in this package can be imported like so:

```js
import { chakraComponents } from "chakra-react-select"
```

The second option would be to completely replace my `Option` component with a custom one.  For the most part, you could just copy [the source for that component](https://github.com/csandman/chakra-react-select/blob/main/src/chakra-components.tsx#L908), modify it to fit your needs, and pass it back into the parent select.

For documentation on how to implement either of those options, you should check out the original [react-select docs on custom components](https://react-select.com/components).

If it wasn't Friday evening I'd throw together a quick example of how you might accomplish this but if you can wait a bit I'll try and get you an example sometime next week!

**Reply by howdyAnkit** [reply]
Went through docs didnt find answer how to set defaultValue can you point or show can that be achieved

**Reply by csandman** [reply]
Sure, you just need to set the `defaultValue` prop on the `Select` component.  From the [original react-select docs](https://react-select.com/props):

> ## [StateManager Props](https://react-select.com/props#statemanager-props)
> The statemanager is a utility class that wraps around the base Select and each Select variant to expose inputValue and value as controllable props. For more detailed information on these props and their usage please see the [controlled props](https://react-select.com/advanced#controlled-props) section of the advanced page.
> 
> `defaultValue`
> ---
> ```
> One of<
>   null,
>   Option,
>   ReadonlyArray<...>
> >
> ```

So you can implement your select with something like this:

```jsx
<Select
  options={options}
  defaultValue={{
    label: "My Default Option",
    value: "my-default-option"
  }}
/>
```

---

However, in a controlled value scenario you'd probably want to just pass in the default value from state:

```jsx
import { useState } from "react";
import { Select } from "chakra-react-select";

const colorOptions = [
  { value: "blue", label: "Blue" },
  { value: "purple", label: "Purple" },
  { value: "red", label: "Red" },
  { value: "orange", label: "Orange" },
  { value: "yellow", label: "Yellow" },
  { value: "green", label: "Green" }
];

const defaultValue = colorOptions[0]

const Example = () => {
  const [value, setValue] = useState(defaultValue);

  return (
    <Select
      name="colors"
      placeholder="Select some colors..."
      value={value}
      options={colorOptions}
      onChange={setValue}
    />
  );
};
```

CodeSandbox Example: https://codesandbox.io/s/default-value-controlled-hzdj69?file=/example.js

---

Or if you're using the `isMulti` prop, you'll want to pass the default value in as an array of options:

```jsx
// ...
const defaultValue = [colorOptions[0]]
// ...
```

**Comment by csandman** [comment, marked as answer]
So I know I said I'd make an example for how to do this before, but I kind of spaced on it for a while.  I did end up making an example of how you could implement this since then though.  In fact I made 2 versions, one using the styles from the Chakra UI `Radio` and one using the `Checkbox`.

## Radio

You example image looks very close to the chakra-ui Radio so I'll start with that.  In order to implement it I, I made a custom `Option` component and pulled the styles for the `Radio` from the Chakra theme.  Here's the code/example image:

CodeSandbox: https://codesandbox.io/s/chakra-react-select-radio-q1vvgg?file=/example.js

<img width="927" alt="image" src="https://user-images.githubusercontent.com/9214195/159059904-f4f4dff9-3282-4d2a-a2cf-f298596f420e.png">

```jsx
import { useMultiStyleConfig, chakra } from "@chakra-ui/react";
import { Select, chakraComponents } from "chakra-react-select";

export const colorOptions = [
  { value: "blue", label: "Blue" },
  { value: "purple", label: "Purple" },
  { value: "red", label: "Red" },
  { value: "orange", label: "Orange" },
  { value: "yellow", label: "Yellow" },
  { value: "green", label: "Green" }
];

const Option = ({ children, isSelected, ...props }) => {
  const { control } = useMultiStyleConfig("Radio", { colorScheme: "orange" });

  // Extra styles added to the radio control
  // https://github.com/chakra-ui/chakra-ui/blob/main/packages/radio/src/radio.tsx#L111
  const css = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    flexShrink: 0,
    ...control,
    ...(isSelected && control._checked)
  };

  return (
    <chakraComponents.Option {...props}>
      <chakra.span __css={css} mr={3} />
      {children}
    </chakraComponents.Option>
  );
};

const Example = () => (
  <Select
    name="colors"
    options={colorOptions}
    placeholder="Select some colors..."
    closeMenuOnSelect={false}
    components={{ Option }}
  />
);
```

## Checkbox

While you didn't request a `Checkbox` styled one, I thought it might be better to use a checkbox type style for a multi select as thats generally what checkboxes are for.  Unfortunately it wasn't as straight forward as the `Radio` because the Chakra checkbox uses a custom `CheckboxIcon` component that they don't export from `@chakra-ui/react`.  So for the sake of example I just copied the source for that component straight into the example, however you could probably also include it by importing it from `@chakra-ui/checkbox`.

CodeSandbox: https://codesandbox.io/s/chakra-react-select-checkbox-uvn325?file=/example.js

<img width="848" alt="image" src="https://user-images.githubusercontent.com/9214195/159061023-00f832f0-ffa0-4225-8e66-2878bb22fbfe.png">

```jsx
import { useMultiStyleConfig, chakra } from "@chakra-ui/react";
import { Select, chakraComponents } from "chakra-react-select";
import { CheckboxIcon } from "./components/checkbox-icon";

export const colorOptions = [
  { value: "blue", label: "Blue" },
  { value: "purple", label: "Purple" },
  { value: "red", label: "Red" },
  { value: "orange", label: "Orange" },
  { value: "yellow", label: "Yellow" },
  { value: "green", label: "Green" }
];

// Based on the source for @chakra-ui/checkbox
// https://github.com/chakra-ui/chakra-ui/blob/main/packages/checkbox/src/checkbox.tsx#L18
const CheckboxControl = chakra("span", {
  baseStyle: {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    verticalAlign: "top",
    userSelect: "none",
    flexShrink: 0
  }
});

const Option = ({ children, isSelected, ...props }) => {
  const { control, icon } = useMultiStyleConfig("Checkbox", {
    colorScheme: "orange"
  });

  // https://github.com/chakra-ui/chakra-ui/blob/main/packages/checkbox/src/checkbox.tsx#L134
  const iconStyles = {
    opacity: isSelected ? 1 : 0,
    transform: isSelected ? "scale(1)" : "scale(0.95)",
    ...icon
  };

  return (
    <chakraComponents.Option {...props}>
      <CheckboxControl
        __css={{ ...control, ...(isSelected && control._checked) }}
        mr={3}
      >
        <CheckboxIcon __css={iconStyles} isChecked={isSelected} />
      </CheckboxControl>
      {children}
    </chakraComponents.Option>
  );
};

const Example = () => (
  <Select
    isMulti
    name="colors"
    options={colorOptions}
    placeholder="Select some colors..."
    closeMenuOnSelect={false}
    hideSelectedOptions={false}
    components={{ Option }}
  />
);
```

**Comment by howdyAnkit** [comment]
Damm Thankyou so much for All the Support and Help. I overlooked this stuff but thanks for getting into details helping me out.

**Reply by csandman** [reply]
No worries, glad I could help!

**Comment by sarates** [comment]
@csandman thanks for these suggestions. But how to change the bg color for all selected options? (`isMulti` enabled). 
I'm trying to adjust `menuItem` styles, but this stuff is working for the last hovered active item :( 
```
const baseStyle = definePartsStyle({
  item: {
    _active: {
      backgroundColor: 'primary.200',
    },
  },
});

export const Menu = defineMultiStyleConfig({ baseStyle });
```

**Reply by csandman** [reply]
Have you looked into using the [`selectedOptionColor`](https://github.com/csandman/chakra-react-select#selectedoptioncolor--default-blue) prop? It will only let you pick a color scheme (not a specific background color) but if that's fine then this is the best course to take.

Otherwise, I'd recommend using [`chakraStyles`](https://github.com/csandman/chakra-react-select#chakrastyles) to accomplish this:

```tsx
return (
  <Select
    isMulti
    hideSelectedOptions={false}
    chakraStyles={{
      option: (provided, { isSelected }) => ({
        ...provided,
        bg: isSelected ? "blue.900" : provided.bg,
      }),
    }}
  />
);
```

CodeSandbox: https://codesandbox.io/s/h3of7z?file=/app.tsx

---

You are talking about the selected in the list, right?  Otherwise, if you're talking about the tags that appear for each selected option, you'd want to use the [`colorScheme`](https://github.com/csandman/chakra-react-select#colorscheme) prop, or apply the `chakraStyles` to `multiValue`.
